File: leetcode/java/OddEvenLinkedList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
public class OddEvenLinkedList {
    public ListNode oddEvenList(ListNode head) {
        if (head != null) {
            ListNode oddPointer = head;
            ListNode evenPointer = head.next;
            ListNode evenHead = head.next;

            while (oddPointer.next != null && oddPointer.next.next != null) {
                ListNode temp = evenPointer.next.next;
                oddPointer.next = oddPointer.next.next;
                evenPointer.next = temp;
                oddPointer = oddPointer.next;
                evenPointer = evenPointer.next;
            }
            oddPointer.next = evenHead;
        }
        return head;
    }
}

File: leetcode/kotlin/odd-even-linked-list.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun oddEvenList(head: ListNode?): ListNode? {
        val start = ListNode(-1)
        val evenHead: ListNode? = head?.next
        var evenTail: ListNode? = start
        var node: ListNode? = head
        var prev: ListNode? = null
        
        while (node != null) {
            evenTail?.next = node.next
            node.next = node.next?.next
            prev = node
            
            evenTail = evenTail?.next
            node = node.next
        }
        prev?.next = evenHead
        return head
    }
}